File: ru-olymp-train-winter-2009/Problems/081221/tree/unit/demo/treeunit.c
#include "treeunit.h"

#include <stdio.h>
#include <stdlib.h>

#define INPUT_FILE "input.txt"
#define MAXN 200000
#define MAXD 5
#define MAXQ 100

struct Edge {
    int v, n;
};

static int initialized = 0, numQ = 0;
static int n, ne, secret;
static int a[MAXN + 1], b[MAXN + 1], p[MAXN + 1], f[MAXN + 1], deg[MAXN + 1], depth[MAXN + 1];
static char buf[1000];
static struct Edge e[2 * MAXN + 1];

void reportError(char *reason) {
    printf("%s\n", reason);
    exit(0);
}

int get(int x) {
    if (x != p[x]) p[x] = get(p[x]);
    return p[x];
}

void unite(int x, int y) {
    x = get(x), y = get(y);
    if (rand() % 2) p[x] = y;
    else p[y] = x;
}

void addEdge(int x, int y) {
    ne++;
    e[ne].v = y;
    e[ne].n = f[x];
    f[x] = ne;    
    ne++;
    e[ne].v = x;
    e[ne].n = f[y];
    f[y] = ne;
}

void dfs(int x, int p, int d) {
    int cur;
    depth[x] = d;
    for (cur = f[x]; cur; cur = e[cur].n) {
        if (e[cur].v == p) {
            continue;
        }
        dfs(e[cur].v, x, d + 1);
    }
}

void init() {
    int i;
    printf("init()\n");
    if (initialized) {
        sprintf(buf, "already initialized");
        reportError(buf);
    }
    srand(0xdead);
    initialized = 1;
    FILE *inf = fopen(INPUT_FILE, "r");
    if (!inf) {
        reportError("input file not found");
    }
    fscanf(inf, "%d", &n);
    if (n < 1 || n > MAXN) {
        sprintf(buf, "invalid n: %d", n);
        reportError(buf);
    }
    for (i = 1; i <= n; i++) {
        p[i] = i;
        f[i] = 0;
        deg[i] = 0;
    }
    ne = 0;
    for (i = 1; i < n; i++) {
        fscanf(inf, "%d%d", &a[i], &b[i]);
        if (a[i] < 1 || a[i] > n) {
            sprintf(buf, "invalid a[%d]: %d", i, a[i]);
            reportError(buf);
        }
        if (b[i] < 1 || b[i] > n) {
            sprintf(buf, "invalid b[%d]: %d", i, b[i]);
            reportError(buf);
        }
        if (get(a[i]) == get(b[i])) {
            sprintf(buf, "invalid graph: not a tree");
            reportError(buf);
        }
        unite(a[i], b[i]);
        addEdge(a[i], b[i]);
        deg[a[i]]++;
        deg[b[i]]++;
    }
    for (i = 1; i <= n; i++) {
        if (deg[i] > MAXD) {
            sprintf(buf, "invalid degree of vertex %d: %d", i, deg[i]);
            reportError(buf);
        }
    }
    fscanf(inf, "%d", &secret);
    if (secret < 1 || secret > n) {
        sprintf(buf, "invalid secret: %d", secret);
        reportError(buf);
    }
    fclose(inf);
    dfs(secret, 0, 0);
}

int getN(void) {
    printf("getN()\n");
    if (!initialized) {
       reportError("not initialized");
    }
    return n;
}

int getA(int edgeNum) {
    printf("getA(%d)\n", edgeNum);
    if (!initialized) {
        reportError("not initialized");
    }
    if (edgeNum < 1 || edgeNum >= n) {
        sprintf(buf, "invalid call: getA(%d)", edgeNum);
        reportError(buf);
    }
    return a[edgeNum];
}

int getB(int edgeNum) {
    printf("getB(%d)\n", edgeNum);
    if (!initialized) {
        reportError("not initialized");
    }
    if (edgeNum < 1 || edgeNum >= n) {
        sprintf(buf, "invalid call: getB(%d)", edgeNum);
        reportError(buf);
    }
    return b[edgeNum];
}

int query(int edgeNum) {
    int i;
    printf("query(%d)\n", edgeNum);
    if (!initialized) {
        reportError("not initialized");
    }
    if (edgeNum < 1 || edgeNum >= n) {
        sprintf(buf, "invalid call: query(%d)", edgeNum);
        reportError(buf);
    }
    numQ++;
    if (numQ > MAXQ) {
        sprintf(buf, "too many queries: %d", numQ);
        reportError(buf);
    }
    return depth[a[edgeNum]] > depth[b[edgeNum]];
}

void report(int vertexNum) {
    printf("report(%d)\n", vertexNum);
    if (!initialized) {
        reportError("not initialized");
    }
    if (vertexNum < 1 || vertexNum > n) {
        sprintf(buf, "invalid call: report(%d)", vertexNum);
        reportError(buf);
    }
    if (vertexNum == secret) {
        printf("ok: %d queries\n", numQ);
    }
    else {
        printf("wa: %d expected but %d found\n", secret, vertexNum);
    }
    exit(0);
}

File: ru-olymp-train-winter-2009/Problems/081221/tree/unit/demo/treeunit.pas
unit
  treeunit;

interface

procedure init;
function getN: longint;
function getA(edgeNum: longint): longint;
function getB(edgeNum: longint): longint;
function query(edgeNum: longint): longint;
procedure report(vertexNum: longint);

implementation

uses
  SysUtils;

const
  inputFile = 'input.txt';
  outputFile = 'output.txt';
  maxN = 200000;
  maxD = 5;
  maxQ = 100;

type
  Edge = record
    v, n: longint;
  end;

var
  inf: text;
  n, i, secret, ne, numQ: longint;
  a, b, p, f, depth, deg: array [1 .. maxN] of longint;
  e: array [1 .. 2 * maxN] of Edge;
  initialized: boolean;

procedure reportError(reason: string);
begin
  writeln(reason);
  halt(0);
end;

procedure init;
begin
  writeln('init()');
  if initialized then
    reportError('already initialized');
  initialized := true;
end;

function getN: longint;
begin
  writeln('getN()');
  if not initialized then
    reportError('not initialized');
  getN := n;
end;

function getA(edgeNum: longint): longint;
begin
  writeln('getA(' + inttostr(edgeNum) + ')');
  if not initialized then
    reportError('not initialized');
  if (edgeNum < 1) or (edgeNum >= n) then
    reportError('invalid call: getA(' + intToStr(edgeNum) + ')');
  getA := a[edgeNum];
end;

function getB(edgeNum: longint): longint;
begin
  writeln('getB(' + inttostr(edgeNum) + ')');
  if not initialized then
    reportError('not initialized');
  if (edgeNum < 1) or (edgeNum >= n) then
    reportError('invalid call: getB(' + intToStr(edgeNum) + ')');
  getB := b[edgeNum];
end;

function query(edgeNum: longint): longint;
begin
  writeln('query(' + inttostr(edgeNum) + ')');
  if not initialized then
    reportError('not initialized');
  if (edgeNum < 1) or (edgeNum >= n) then
    reportError('invalid call: query(' + intToStr(edgeNum) + ')');
  inc(numQ);
  if numQ > maxQ then
    reportError('too many queries: ' + inttostr(numQ));
  if (depth[a[edgeNum]] < depth[b[edgeNum]]) then
    query := 0
  else query := 1;
end;

procedure report(vertexNum: longint);
begin
  writeln('report(' + inttostr(vertexNum) + ')');
  if not initialized then
    reportError('not initialized');
  if (vertexNum < 1) or (vertexNum > n) then
    reportError('invalid call: report(' + intToStr(vertexNum) + ')');
  if (vertexNum = secret) then
    writeln('ok: ' + inttostr(numQ) + ' queries')
  else
    writeln('wa: ' + inttostr(secret) + ' expected, but ' + inttostr(vertexNum) + ' found');
  halt(0);
end;

function get(x: longint): longint;
begin
  if (x <> p[x]) then
    p[x] := get(p[x]);
  get := p[x];
end;

procedure unite(x, y: longint);
begin
  x := get(x);
  y := get(y);
  if (random(2) = 0) then
    p[x] := y
  else
    p[y] := x;
end;

procedure addEdge(a, b: longint);
begin
  inc(ne);
  e[ne].v := b;
  e[ne].n := f[a];
  f[a] := ne;
  inc(ne);
  e[ne].v := a;
  e[ne].n := f[b];
  f[b] := ne;
end;

procedure dfs(x, p, d: longint);
var
  cur: longint;
begin
  depth[x] := d;
  cur := f[x];
  while cur <> 0 do
  begin
    if e[cur].v <> p then
      dfs(e[cur].v, x, d + 1);
    cur := e[cur].n;
  end;
end;

begin
  initialized := false;
  randseed := $dead;
  {$I-}
  assign(inf, inputFile);
  reset(inf);
  if ioresult <> 0 then
    reportError('input file not found');
  read(inf, n);
  if (n < 1) or (n > maxN) then
    reportError('invalid n: ' + inttostr(n));
  for i := 1 to n do
  begin
    p[i] := i;
    f[i] := 0;
    deg[i] := 0;
  end;
  for i := 1 to n - 1 do
  begin
    read(inf, a[i], b[i]);
    if (a[i] < 1) or (a[i] > n) then
      reportError('invalid a[' + inttostr(i) + ']: ' + inttostr(a[i]));
    if (b[i] < 1) or (b[i] > n) then
      reportError('invalid b[' + inttostr(i) + ']: ' + inttostr(b[i]));
    if (get(a[i]) = get(b[i])) then
      reportError('invalid graph: not a tree');
    unite(a[i], b[i]);
    addEdge(a[i], b[i]);
    inc(deg[a[i]]);
    inc(deg[b[i]]);
  end;
  for i := 1 to n do
  begin
    if deg[i] > maxD then
      reportError('invalid degree of vertex ' + inttostr(i) + ': ' + inttostr(deg[i]));
  end;
  read(inf, secret);
  if (secret < 1) or (secret > n) then
    reportError('invalid secret: ' + inttostr(secret));
  close(inf);
  dfs(secret, 0, 0);
  numQ := 0;
end.
